### PR TITLE
chore: migrate s3 ops to sdk v2, add workflow for loading sdk v2 config

### PR DESCRIPTION
**Pull request by lucix-aws**
Part of #187

I identified this project as an advanced user of the Go SDK for the following reasons:
* The nature of the project requires calling _many_ AWS services
* The way the project loads config is sufficiently more advanced than the more common / basic use cases

I did this migration as an exercise to get a better idea for what doing so is like in a large project and to see if there's anything we can improve in that area. This made me realize that an outstanding issue we have https://github.com/aws/aws-sdk-go-v2/issues/2422 is more important of a parity gap than I thought, its priority has been increased accordingly.

What this PR does:
* establishes the workflow for loading SDK v2 `aws.Config`, which is the v2 equivalent of `session.Session`.
* migrates all S3 resources to using SDK v2
  * The actual call-to-call translation from v1 to v2 is pretty straightforward. Note that I switched use of ListObjects to ListObjectsV2 because the latter has a paginator available. It shouldn't affect anything behaviorally that I'm aware of.
* The libnuke facade gives you a `ctx` everywhere for doing most/all service calls, but `config.LoadDefaultConfig` also takes one, in places where I didn't have access I just passed `context.TODO()`
* I don't know what your stance is on stability of exported APIs in this project, some of them were probably broken by this change.

What I tested:
* Ran the existing s3-bucket integration test. Note that I'm getting a failure from running this against both main and this branch, so I assume that's existing. The actual nuke command also worked so I didn't investigate it that much.
* Actually ran nuke against my dev account to blow up a bunch of old testing buckets:
```
regions:
  - us-east-1

blocklist:
  - "999999999999" # production

accounts:
  "<...>": 
    filters:
      S3Bucket:
        # ... filtered a bunch of stuff I wanted to keep

resource-types:
  includes:
    - S3Bucket
```
* The "custom service" / endpoint URL path probably warrants some additional e2e testing. I don't have a good setup for this.

You definitely _can_ have both SDK v1 and v2 live alongside each other, but you may or may not wish to cut releases in that transition period. If that's the case I would just change this to merge to a feature branch.

### CODE REVIEW

File: pkg/awsutil/config.go [hunk 1]
```diff
@@ -0,0 +1,199 @@
...
+
+		static := credentials.NewStaticCredentialsProvider(
+			creds.AccessKeyID,
+			creds.SecretAccessKey,
+			creds.SessionToken,
+		)
+		opts = append(opts, config.WithCredentialsProvider(static))
+
+	case c.HasProfile() && c.HasKeys():
+		return nil, fmt.Errorf("you have to specify a profile or credentials for at least one region")
```
Comment (wty-Bryant) [inline pkg/awsutil/config.go:113]: nit: might be better to say "can not specify both profile and credentials for one region" considering it is an error message to correct user config behavior

File: pkg/awsutil/config.go [hunk 2]
```diff
@@ -0,0 +1,199 @@
...
+	return "aws-nuke::traceRequest"
+}
+
+func (traceRequest) HandleFinalize(
+	ctx context.Context, in middleware.FinalizeInput, next middleware.FinalizeHandler,
+) (
+	middleware.FinalizeOutput, middleware.Metadata, error,
+) {
+	req, ok := in.Request.(*smithyhttp.Request)
+	if !ok {
```
Comment (wty-Bryant) [inline pkg/awsutil/config.go:165]: can be simplified using the same logic of `traceResponse`:
```
if ok {
    log.Tracef("sending AWS request:\n%s", DumpRequest(req.Request))
}
return next.HandleFinalize()
```

File: pkg/awsutil/config.go [hunk 3]
```diff
@@ -0,0 +1,199 @@
...
+func (traceResponse) HandleDeserialize(
+	ctx context.Context, in middleware.DeserializeInput, next middleware.DeserializeHandler,
+) (
+	middleware.DeserializeOutput, middleware.Metadata, error,
+) {
+	out, md, err := next.HandleDeserialize(ctx, in)
+
+	resp, ok := out.RawResponse.(*smithyhttp.Response)
+	if ok {
+		log.Tracef("sending AWS request:\n%s", DumpResponse(resp.Response))
```
Comment (wty-Bryant) [inline pkg/awsutil/config.go:188]: receiving AWS response?

File: pkg/awsutil/session.go [hunk 4]
```diff
@@ -47,6 +51,7 @@ type Credentials struct {
 
 	CustomEndpoints config.CustomEndpoints
 	session         *session.Session
```
Comment (wty-Bryant) [inline pkg/awsutil/session.go:53]: one question not impacting the migration: we replace `DefaultRegionID`/`DefaultAWSPartitionID` and other API in batch.go with go v2's value and code, could we directly remove `session` field here considering it is v1 only?

**Comment by ekristen** [comment]
@lucix-aws This is amazing. It was on my list to look at doing this in earnest this winter, but this looks like a great start.

It has been a while since I've run the integration tests, so it's possible they are broken, I will take a look.

Generally not worried about exported APIs in this project, it's not meant to be directly used, and if it is it's usually at a higher level that making changes to the individual resources shouldn't be a problem. 

I hadn't yet decided if I was going to run them in parallel or not. Ultimately I think I might have to as it's quite the lift to shift everything over at once especially with so many changes. 

I will definitely take a closer look at this and this will help me form the strategy for migrating.

**Review by wty-Bryant — APPROVED** [review]
nice example to learn migration in detail, left some minor comments

**Comment by ekristen** [comment]
[This fix](https://github.com/ekristen/aws-nuke/pull/381) fixes the integration tests, you'll likely need to rebase.

File: pkg/awsutil/config.go (hunk 1 above)
Comment (lucix-aws) [reply pkg/awsutil/config.go:113]: This string was taken from the v1 version of this so I'm going to leave as-is.

**Comment by lucix-aws** [comment]
Patch addresses PR comments, lint, and rebases for the integration tests.

**Comment by lucix-aws** [comment]
I PR'd the addition of `config.WithBaseEndpoint` SDK-side and will update once that releases. I'll also deal w/ the commit message linting then.

**Comment by mdgm88** [comment]
Upgrading s3 to using the sdk v2 could also be handy for using list buckets. It's now possible to filter list buckets on a region.

**Comment by lucix-aws** [comment]
We are mid-hackathon in SDKs and Tools right now, will get back to this next week.

**Comment by ekristen** [comment]
No problem @lucix-aws. Looking forward to when this can be merged and we can the ball rolling on moving to sdk v2 for all resources.

**Comment by lucix-aws** [comment]
Fixed commits and pushed an update to use the new `config.WithBaseEndpoint`. This should be ready for review/merge.

**Review by ekristen — COMMENTED** [review]
Looks great @lucix-aws I really appreciate this, going to save me a lot of time. Looks like we have a small merge conflict on go.mod though. I'll start testing it locally in the mean time.

**Comment by ekristen** [comment]
@lucix-aws I've been testing through this, I found one small problem with the integration tests and I fixed that. 

However it does look like we are also missing parity from the session function to the config function around skipping the global handler https://github.com/ekristen/aws-nuke/blob/main/pkg/awsutil/session.go#L207C1-L210C3

**Comment by ekristen** [comment]
```> aws-nuke - 3.0.0-dev - dirty
Do you really want to nuke the account with the ID 637423565883 and the alias 'no-alias-637423565883'?
Waiting 3s before continuing.
global - S3Bucket - s3://fsdfhasdjfohsgdalfha-989908790-89080890890890898908 - [CreationDate: "2024-11-01T22:49:00Z", Name: "fsdfhasdjfohsgdalfha-989908790-89080890890890898908", ObjectLock: ""] - would remove
us-east-1 - S3Bucket - s3://fsdfhasdjfohsgdalfha-989908790-89080890890890898908 - [CreationDate: "2024-11-01T22:49:00Z", Name: "fsdfhasdjfohsgdalfha-989908790-89080890890890898908", ObjectLock: ""] - would remove
```

it should be 

```> aws-nuke - 3.0.0-dev - dirty
Do you really want to nuke the account with the ID 637423565883 and the alias 'no-alias-637423565883'?
Waiting 3s before continuing.
us-east-1 - S3Bucket - s3://fsdfhasdjfohsgdalfha-989908790-89080890890890898908 - [CreationDate: "2024-11-01T22:49:00Z", Name: "fsdfhasdjfohsgdalfha-989908790-89080890890890898908", ObjectLock: ""] - would remove
```

**Comment by ekristen** [comment]
https://github.com/aws/aws-sdk-go-v2/issues/2886

**Comment by ekristen** [comment]
I appreciate the feedback, `global` is essentially `us-east-1` except that it's for the things are are truly global, for example IAM users, roles, policies, and a few other select things. I think that I might have to just add a metadata property for "global" resources.

The reason why this exists is because yes you can query IAM in europe or asia, but it's all based out of us-east-1 and we don't want to try and remove it from two locations when it's really just in us-east-1, so the global exists as a way of ensuring we only query IAM stuff once.

**Comment by lucix-aws** [comment]
> I think that I might have to just add a metadata property for "global" resources.

This seems like the right way forward here - that seems like it's really the level you want to be selecting on here (resources), rather than just a particular service.

So, in terms of this PR, do we want to just put a blanket "skip when global" middleware or leave it for now?

**Comment by ekristen** [comment]
An example middleware would be great, I'll need to fiddle with adding the global metadata and figuring out how to tie to together to make it work.

**Comment by lucix-aws** [comment]
done, see https://github.com/ekristen/aws-nuke/pull/376/commits/b7892be9cc63531fd9475f9426a24dc597e24a06

**Comment by ekristen** [comment]
@lucix-aws thanks, this is very helpful.

**Comment by ekristen** [comment]
@lucix-aws any chance you can rebase? I've tested the changes, I'm happy to move forward and get this merged and start putting the project on the path to migrating all the way to v2. Thanks!

**Comment by lucix-aws** [comment]
Will do this week

**Comment by lucix-aws** [comment]
rebased

**Pull request merged**